### PR TITLE
[WIP] HIVE-21456: addendum: Improvements in Thrift over HTTP for Hive MetaStore

### PR DESCRIPTION
**Pull request by sourabh912**
### What changes were proposed in this pull request?

This PR is a follow up of HIVE-21456. In our internal testing, we found the following issues and this PR fixes them:

- Http thread pool name was not getting set in Jetty server
- Found an issue in which PersistenceManager object was not getting cleaned up after the completion of a http request. Fixed it by adding a listener in Jetty's connector. The listener listens for the completion of Http request and then closes the RawStore (a thread local entity). The RawStore, in turn, cleans up PersistenceManager. 
- Added jetty's dependency in metastore pom.xml. Earlier this dependency was getting honored from hadoop-common.
- Fixed a unit test  testJDOPersistanceManagerCleanup() in TestHiveMetaStore

### Why are the changes needed?

To fix issues described above

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Fixed an existing test i.e testJDOPersistanceManagerCleanup. The failure in the test was unrelated to thrift over http patch. After the test is fixed, it was failing for HiveMetaStore in HTTP mode. This patch address that

### CODE REVIEW

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java [hunk 1]
```diff
@@ -63,10 +63,14 @@
 
 import org.eclipse.jetty.security.ConstraintMapping;
 import org.eclipse.jetty.security.ConstraintSecurityHandler;
+import org.eclipse.jetty.server.HttpChannel;
 import org.eclipse.jetty.server.HttpConfiguration;
 import org.eclipse.jetty.server.HttpConnectionFactory;
+import org.eclipse.jetty.server.Request;
 import org.eclipse.jetty.server.Server;
 import org.eclipse.jetty.server.ServerConnector;
+
```
Comment (hsnusonic) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java:72]: nit: additional empty line?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java [hunk 2]
```diff
@@ -461,6 +459,14 @@ public Thread newThread(@NotNull Runnable r) {
     // TODO: AcceptQueueSize needs to be higher for HMS
     connector.setAcceptQueueSize(maxWorkerThreads);
     // TODO: Connection keepalive configuration?
+    connector.addBean(new HttpChannel.Listener() {
+      @Override
+      public void onComplete(Request request) {
+        LOG.debug("Request " + request + " is completed. HttpChannel was: " + request.getHttpChannel()
+            + ", HttpTransport: " + request.getHttpChannel().getHttpTransport());
+        HMSHandler.cleanupHandlerContext();
```
Comment (hsnusonic) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java:467]: Is it possible to add a unit test to make sure we indeed clean up all of objects that need to be removed?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java [hunk 3]
```diff
@@ -399,19 +403,13 @@ private static ThriftServer startHttpMetastore(int port, Configuration conf)
...
-        new SynchronousQueue<>(), new ThreadFactory() {
-      @Override
-      public Thread newThread(@NotNull Runnable r) {
-        Thread newThread = new Thread(r);
-        newThread.setName(threadPoolNamePrefix + ": Thread-" + newThread.getId());
-        return newThread;
-      }
-    });
-    ExecutorThreadPool threadPool = new ExecutorThreadPool((ThreadPoolExecutor) executorService);
+    ExecutorThreadPool executorThreadPool = new ExecutorThreadPool(maxWorkerThreads, minWorkerThreads,
```
Comment (hsnusonic) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java:406]: Could you give me some context why the change is needed? It seems netty has default idle timeout 30 seconds, but it was declared 60 seconds explicitly.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java (hunk 1 above)
Comment (sourabh912) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java:72]: Sure, will remove it.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java (hunk 2 above)
Comment (sourabh912) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java:467]: `testJDOPersistanceManagerCleanup` in TestHiveMetaStore is testing that. The failure in that test led to the fixes proposed in this patch. Without this patch, testJDOPersistanceManagerCleanup() fails in Thrift over HTTP mode.

Comment (dengzhhu653) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java:467]: should we track the open connection metric as well?

File: standalone-metastore/pom.xml [hunk 4]
```diff
@@ -432,6 +443,31 @@
...
+      <artifactId>jetty-http</artifactId>
+      <version>${jetty.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>org.eclipse.jetty</groupId>
+      <artifactId>jetty-rewrite</artifactId>
+      <version>${jetty.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>org.eclipse.jetty</groupId>
```
Comment (dengzhhu653) [inline standalone-metastore/pom.xml:457]: Do we only need jetty-server and jetty-servlet for HMS currently?

Comment (sourabh912) [reply standalone-metastore/pom.xml]: Looks like yes. I will take a look and remove dependencies which are not required.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java (hunk 2 above)
Comment (sourabh912) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java]: @dengzhhu653 : Yes we should. Since that is unrelated to this patch, can we take it up in a separate task ?

Comment (dengzhhu653) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java]: My idea is that this is a relative minor change, we can do it as a plus, igore this if you already have a plan for the metric.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java [hunk 5]
```diff
@@ -18,9 +18,11 @@
 
 package org.apache.hadoop.hive.metastore;
 
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
```
Comment (dengzhhu653) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java:21]: some imports can be cleared.

**Review by dengzhhu653 — APPROVED** [review]

**Comment by dengzhhu653** [comment]
Leave a minor comment, others look good to me, +1.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java [hunk 6]
```diff
@@ -398,18 +404,16 @@ private static ThriftServer startHttpMetastore(int port, Configuration conf)
...
-        newThread.setName(threadPoolNamePrefix + ": Thread-" + newThread.getId());
-        return newThread;
-      }
+    final String threadPoolNameFormat = "Metastore-HTTPHandler-Pool: Thread-%s";
+    // TODO: Add a config for keepAlive time of threads ?
+    ExecutorService executorService = new ThreadPoolExecutor(minWorkerThreads, maxWorkerThreads, 60L,
+        TimeUnit.SECONDS, new SynchronousQueue<>(), r -> {
+      Thread thread = new Thread(r);
+      thread.setDaemon(true);
+      thread.setName("Metastore-HttpHandler-Pool: Thread-" + thread.getId());
```
Comment (nrg4878) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java:413]: nit: should you use the "threadPoolNameFormat" variable above?

Comment (sourabh912) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java:413]: The variable is no longer required in the current implementation. Will remove it.

**Review by nrg4878 — APPROVED** [review]
